Experimental pull request

rneswold opened the pull request:
This uses a recently added feature of Dart.

It wraps type information around a Dart primitive -- in this case, an `int` -- to restrict how it is used. We used an `int` as a plot configuration ID. Unfortunately, a programmer could accidentally use it anywhere an `int` is allowed. A class is too heavy-handed to limit an `int`'s use. Dart's new "extension type" allows us to pass around an `int`, but can't use it in expressions. We do allow it to be compared with other config IDs, so they can be used as keys in a map.

This is an experiment, so I'd like to test it with flutter apps by specifying this `pull-request` branch before we merge it for good.

rneswold committed:
:recycle: clean-up imports

rneswold committed:
:sparkles: make plot config ID an extension type

This wraps the `int` with a type context so you can't do "int-y"
things with it (like add two IDs together.) It does provide the
comparison API so two IDs can be compared.

rneswold committed:
:label: v0.2.9

beauremus commented:
So, it treats the extension as a unique type with the underlying implementation still matching the extended type?
WHOA! Just read the docs on this. The ability to constrain the type is awesome 😲 
https://dart.dev/language/extension-types

beauremus commented on lib/src/service/acsys_service.dart:
```diff
@@ -370,8 +369,18 @@ class ChannelSettingSnapshot {
+/// supports comparisons.
+
+extension type PlotConfigId._(int raw) implements Comparable {
+  PlotConfigId._fromInt(this.raw);
```
Why do you need `this.raw` here when in other places it's referred to with `raw`?

rneswold replied:
`PlotConfigId._fromInt` is a named constructor for the type (it also happens to be private to the module.) The `this.raw` is Dart notation that the argument should be placed in the type's field.

The rest of the access is assumed to be in the context of the type (because there isn't a `static` keyword in front), so `raw` is taken to be the internal value.

rneswold replied:
By the way, had I used this:

```
extension type PlotConfigId(int raw) implements Comparable {
```

the compiler would have made a constructor `PlotConfigId(int raw)`. I don't want people making IDs on their own and only want the named constructor to do it. Adding the `._` after the type name is a Dart indicator that the default constructor is private.

beauremus commented on lib/src/service/acsys_service.dart:
```diff
@@ -929,14 +938,16 @@ class ACSysService implements ACSysServiceAPI {
         xlat: (GPlotConfigsData data) => data.plotConfiguration
             .map((e) => PlotConfigurationListing(
-                configurationId: e.configurationId,
+                configurationId: PlotConfigId._fromInt(e.configurationId),
```
Why does the `_fromInt` method have an underscore? I thought the underscore symbolized use constrained to the implementation of the parent scope, `PlotConfigId` in this case, when we're in `PlotConfigurationListing`.

beauremus replied:
I'll answer my own question.
https://stackoverflow.com/a/53144953/2271815
> It's not just a naming convention. Underscore fields, classes and methods will only be available in the .dart file where they are defined.

So, the scope is the .dart file, not the class or block scope.

rneswold replied:
Right. The module needs access to those methods to send the `int` value over the network but any application using the code can't create their own arbitrary IDs. They can only get valid IDs from the GraphQL API.

rneswold replied:
In Dart, functions and variables with a leading underscore are private.

beauremus reviewed (commented):
Looks good to me! I mostly have questions of ignorance. Cool new feature!

rneswold commented:
> So, it treats the extension as a unique type with the underlying implementation still matching the extended type? WHOA! Just read the docs on this. The ability to constrain the type is awesome 😲 https://dart.dev/language/extension-types

It seems to me that lots of developers think of programming as just specifying the steps of an algorithm. But modern languages have features that let you model your problem, which constrains the steps you're able to make and guides you to a solution.

cnlklink commented:
> It seems to me that lots of developers think of programming as just specifying the steps of an algorithm. But modern languages have features that let you model your problem, which constrains the steps you're able to make and guides you to a solution

💯this - and it lets you be clear to other developers on the intent of something like `PlotConfigurationId`.  E.g. "this thing is meant to be generated once by a process you shouldn't care about and then you shouldn't do anything with it after that".

I'l grab this branch and test it against `plotting-app`.

beauremus commented:
> lets you be clear to other developers on the intent of something

This is HUGE

Let's you codify the intent, not just a comment.

cnlklink commented:
As predicted, we'll need a way to generate a sequence of `PlotConfigId`s for the `FakeAcSysService` in `plotting-app`.  Once that's in place I think we can merge!

rneswold commented:
Do you have a function signature in mind? Do we just want a function that returns the next available? Or do you want one that returns a sequence of them?

A function that accepts an arbitrary integer isn't great because then the tests could violate a contract (in that the system won't generate the same ID for two different configurations.)

cnlklink commented:
I was thinking `PlotConfigId.generate()` justs spits out a new / unique ID.  Doesn't have to be a sequence, just has to be unique.

rneswold committed:
:sparkles: add `PlotConfigId.generate()`

This constructor should only be used by unit tests. Applications
shouldn't create their own IDs.

rneswold committed:
:recycle: mark classes `final`

These classes are meant to be the leaf nodes of a class hierarchy.

rneswold committed:
:recycle: make `Status` an extension type

cnlklink reviewed (approved):
Tested and confirmed that this works with `plotting-app` and `FakeAcsysService`.
